posts/2022-01-21-linking-metrics-and-traces-with-exemplars.adoc

```diff
@@ -0,0 +1,154 @@
+      .register()
+----
+Prometheus supports exemplars for histograms and counter metrics.
+
+
+For the tracing we are using OpenTelemetry. Here this is the code to start a single span.
+
+[source,scala]
+----
+tracer
+  .spanBuilder(operationName)
+  .setParent(Context.current())
+  .startSpan()
+----
+
+In this example the trace will contain a single span, just for demonstration purposes.
+
+=== Setup a docker-compose
+In order to create the Exemplars visualization a few components are needed:
+
+- Prometheus
+- OpenTelemetry Collector
+- Grafana
+- Grafana Tempo
+
+The complete setup can be found in https://github.com/lunatech-labs/lunatech-traces-metrics-exemplars/tree/master/docker[the docker folder]. I will only point out a few important details needed to bring Examplars together.
+
+==== Prometheus
+The following command has to be passed in the docker setup
+[source,text]
+----
+--enable-feature=exemplar-storage
+----
+
+==== Grafana
+Note the use of the `TraceID` tag in the setup of the https://github.com/lunatech-labs/lunatech-traces-metrics-exemplars/blob/master/docker/grafana/provisioning/datasources/datasource.yml[datasource] from Prometheus (also used in the source code):
+[source,text]
+----
+ exemplarTraceIdDestinations:
+   - name: TraceID
+     datasourceUid: tempo
+----
+
+
+=== Link docker instances with the candy factory
+Now that we have setup Prometheus in port 13798, we can start an HTTPServer that exports metrics.
+[source,scala]
+----
+val prometheusPort = 13798
+  new HTTPServer(prometheusPort)
+  DefaultExports.initialize()
+----
+
+And finally, we need to also send the traces to the otel-collector:
+[source,scala]
+----
+ val collectorEndpoint = "http://otel-collector:4317"
+  val tracer: Tracer = Tracing.getTracer(collectorEndpoint)
+----
+
+We are still missing the code to instantiate the Tracer. For the sake of reducing the amount of code dumping check here the https://github.com/lunatech-labs/lunatech-traces-metrics-exemplars/blob/master/src/main/scala/com/lunatech/exemplars/observability/Tracing.scala[object Tracing].
+
+=== Grafana dashboard
+In order to run the docker-compose don't forget to first create the `lunatech-traces-metrics-exemplars` image with:
+[source,text]
+----
+$ sbt docker
+----
+Now we are ready to launch the docker-compose:
+[source,text]
+----
+$ docker-compose -f docker/docker-compose.yml up
+----
+
+In the browser Grafana will be available at `localhost:3000`. There's a dashboard already available. If you hover over the little diamonds you will see the exemplar's data details:
+
+.Dashboard with details of one exemplar
+image::dashboard_w_exemplar.png[Exemplar, width = 800]
+
+Clicking on `Query with tempo` will take you to the corresponding trace in Tempo:
+
+.Trace in Tempo
+image::tempo_trace.png[Exemplar, width = 800]
+
+=== Summary
+In this post I have walked you through the steps needed to produce and visualize metrics and traces linked by Exemplars, in a docker-compose setup.
```

Comment:
Maybe adding another sentence here would be nice? Something as simple as "I hope this helps you in your future metrics setups!" or something like that?

Reply:
What about adding `I hope this post has helped you to increase the observability of your systems.`